Content
-------

* `DNAmethylation_genome.Rdata` `DNAmethylation_PGCLC_enhancer.Rdata`: Metadata with the DNA methylation levels: genome-wide and at PGCLC enhancers
* `DNAmethylation_heterogeneity_scNMT.R`: Script for the mCpG methylation heterogeneity
* `Dissimilarity_matrices.R`: Claculation of the Dissimilarity matrices with [PDclust](https://github.com/hui-tony-zk/PDclust)
* `GEO_id_met.txt`: Linkage of the GEO id with the meta data file
* `GSE121650_rna_counts.tsv`: single-cell RNAseq count matrix of the scNMT data
* `diss_scM_Epiblast_PGCLC.Rdata` `diss_scM_Epiblast_EpiLC.Rdata` `diss_scM_Epiblast_genome.Rdata`: Dissimilarity matrix for the PGCLC/EpiLC or genome-wide (output of `Dissimilarity_matrices.R`)
* `sample_metadata.txt`: Meta data file for the [scNMT data](https://github.com/rargelaguet/scnmt_gastrulation)
* `scM_E4_1kb.tab` `scM_E4_1kb.tab` `scM_E4_1kb.tab`: single-cell methylation quantification within 1kb of all PGCLC enhancers for the indicated stages
* `scM_scN_acc_1kb.tab`: single-cell chromatin accessibility (mGpC) within 1kb for all PGCLC enhancers
* `scPGC.Rdata`: single-cell RNAseq expression for all PGCLC genes
* `scRNAseq_analysis.R`: single-cell RNAseq analysis of the scNMT data 
